clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java

```diff
@@ -16,197 +16,78 @@
-            this.networkClientDelegate = new NetworkClientDelegate(
-                    this.time,
-                    this.config,
-                    logContext,
-                    networkClient);
-            this.running = true;
-            this.errorEventHandler = new ErrorEventHandler(this.backgroundEventQueue);
-            this.groupState = new GroupState(rebalanceConfig);
-            long retryBackoffMs = config.getLong(ConsumerConfig.RETRY_BACKOFF_MS_CONFIG);
-            long retryBackoffMaxMs = config.getLong(ConsumerConfig.RETRY_BACKOFF_MAX_MS_CONFIG);
-            final int requestTimeoutMs = config.getInt(ConsumerConfig.REQUEST_TIMEOUT_MS_CONFIG);
-
-            OffsetsRequestManager offsetsRequestManager =
-                    new OffsetsRequestManager(
-                            subscriptionState,
-                            metadata,
-                            configuredIsolationLevel(config),
-                            time,
-                            retryBackoffMs,
-                            requestTimeoutMs,
-                            apiVersions,
-                            networkClientDelegate,
-                            logContext);
-            CoordinatorRequestManager coordinatorRequestManager = null;
-            CommitRequestManager commitRequestManager = null;
-
-            if (groupState.groupId != null) {
-                coordinatorRequestManager = new CoordinatorRequestManager(
-                        this.time,
-                        logContext,
-                        retryBackoffMs,
-                        retryBackoffMaxMs,
-                        this.errorEventHandler,
-                        groupState.groupId);
-                commitRequestManager = new CommitRequestManager(
-                        this.time,
-                        logContext,
-                        subscriptionState,
-                        config,
-                        coordinatorRequestManager,
-                        groupState);
-            }
-
-            this.requestManagers = new RequestManagers(
-                    offsetsRequestManager,
-                    Optional.ofNullable(coordinatorRequestManager),
-                    Optional.ofNullable(commitRequestManager));
-
-            this.applicationEventProcessor = new ApplicationEventProcessor(
-                    backgroundEventQueue,
-                    requestManagers,
-                    metadata);
-
-        } catch (final Exception e) {
-            close();
-            throw new KafkaException("Failed to construct background processor", e.getCause());
-        }
+        this.applicationEventProcessorSupplier = applicationEventProcessorSupplier;
+        this.networkClientDelegateSupplier = networkClientDelegateSupplier;
+        this.requestManagersSupplier = requestManagersSupplier;
     }
 
     @Override
     public void run() {
+        closer.assertOpen("Consumer background thread is already closed");
```

Comment:
I find this name combination closer+assertOpen kind of confusing. I even think it would it be easier to follow if we just had the explicit check `if(isClose()) throw` here

Reply:
Hmm... I can see your point. The naming is definitely confusing 😆

It used to be named `maybeThrowIllegalStateException()`, so it could be worse.